aea/runtime.py

```diff
@@ -0,0 +1,238 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the implementation of runtime for economic agent (AEA)."""
+import asyncio
+import logging
+import threading
+from abc import ABC, abstractmethod
+from asyncio.events import AbstractEventLoop
+from enum import Enum
+from typing import Optional
+
+from aea.agent_loop import AsyncState
+from aea.helpers.async_utils import ensure_loop
+from aea.mail.base import AsyncMultiplexer
+
+if False:
+    # for mypy
+    from aea.agent import Agent
+
+
+logger = logging.getLogger(__name__)
+
+
+class RuntimeStates(Enum):
+    """Runtime states."""
+
+    initial = "not_started"
+    starting = "starting"
+    started = "started"
+    stopped = "stopped"
+
+
+class BaseRuntime(ABC):
+    """Abstract runtime class to create implementations."""
+
+    def __init__(
+        self, agent: "Agent", loop: Optional[AbstractEventLoop] = None
+    ) -> None:
+        """
+        Init runtime.
+
+        :param agent: Agent to run.
+        :param loop: optional event loop. if not provided a new one will be created.
+        :return: None
+        """
+        self._agent: "Agent" = agent
+        self._loop = ensure_loop(
+            loop
+        )  # TODO: decide who constructs loop: agent, runtime, multiplexer.
+        self._state: AsyncState = AsyncState(RuntimeStates.initial)
+
+    def start(self) -> None:
+        """Start agent using runtime."""
+        if self._state.get() is RuntimeStates.started:
+            logger.error("Already started!")
+            return
+        self._start()
+
+    def stop(self) -> None:
+        """Stop agent and runtime."""
+        logger.debug("Runtime stop called!")
+        self.teardown()
+        self._stop()
+        logger.debug("[{}]: Stopped".format(self._agent.name))
+
+    def teardown(self) -> None:
+        """Tear down agent."""
+        logger.debug("Runtime teardown ...")
+        self._agent.teardown()
+        logger.debug("[{}]: Teardown completed".format(self._agent.name))
+
+    @abstractmethod
+    def _start(self) -> None:
+        """Implement runtime start function here."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def _stop(self) -> None:
```

Comment:
Not sure why they are labelled as private.. Either all private or none

Reply:
nice point!